File: components/LoadAdmissions/LeakyBucketForGateway.py
import json


class LeakyBucketForGateway:
    def __init__(self, device_id, tokens, capacities, admission_rate):
        self.device_id = device_id
        self.tokens = tokens
        self.capacities = capacities
        self.admission_rate = admission_rate
        self.leakybucket_history_file = []

        self.last_minute = 0
        self.last_time = 0.0
        self.admission_num = capacities.copy()
        self.init_admission_num()

    @classmethod
    def from_json(cls, json_data):
        """
        从 JSON 字符串或 JSON 文件内容初始化实例。
        参数：
            - json_data：JSON 格式的字符串或字典对象。
        """
        if isinstance(json_data, str):
            # 将 JSON 字符串转换为字典
            data = json.loads(json_data)
        elif isinstance(json_data, dict):
            # 直接使用字典
            data = json_data
        else:
            raise ValueError("Invalid JSON data. Must be a JSON string or dictionary.")

        # 转换 JSON 中的字符串键为元组键
        tokens = {tuple(map(int, k.strip("()").split(", "))): v for k, v in data["tokens"].items()}
        capacities = {tuple(map(int, k.strip("()").split(", "))): v for k, v in data["capacities"].items()}
        admission_rate = {tuple(map(int, k.strip("()").split(", "))): v for k, v in data["admission_rate"].items()}

        return cls(
            device_id=data["device_id"],
            tokens=tokens,
            capacities=capacities,
            admission_rate=admission_rate
        )

    def to_json(self):
        """
        将实例转换为 JSON 格式的字符串。
        """
        # 将元组键转换为字符串
        tokens = {str(k): v for k, v in self.tokens.items()}
        capacities = {str(k): v for k, v in self.capacities.items()}
        admission_rate = {str(k): v for k, v in self.admission_rate.items()}

        # 构建 JSON 数据
        data = {
            "device_id": self.device_id,
            "tokens": tokens,
            "capacities": capacities,
            "admission_rate": admission_rate
        }

        # 转换为 JSON 字符串并返回
        return json.dumps(data, indent=4)

    def init_admission_num(self):
        for key in self.admission_num.keys():
            self.admission_num[key] = 0

    def write_admission_executor_result(self, this_minute):
        for key, num in self.admission_num.items():
            self.leakybucket_history_file.append([this_minute, self.device_id, key[0], key[1], num])

    def update_tokens(self, simulator_time):
        current_time = simulator_time
        for key in self.tokens.keys():
            if key in self.admission_rate and key in self.capacities:
                time_diff = (current_time - self.last_time) / 60.0
                new_tokens = min(self.capacities[key], self.tokens[key] + time_diff * self.admission_rate[key])
                self.tokens[key] = new_tokens
            else:
                print("Error - Cannot find keys for tokens or capacities.")
        self.last_time = current_time

    def is_admission(self, network_packet, simulator_time):
        if network_packet['destination_service_id'] != 0:
            return True
        else:
            this_minute = int(simulator_time / 60)
            if this_minute != self.last_minute:
                self.write_admission_executor_result(this_minute)
                self.init_admission_num()
                self.last_minute = this_minute

            self.update_tokens(simulator_time)
            service_chain_id = network_packet['service_chain_id']
            user_level = network_packet['user_level']
            key = (service_chain_id, user_level)

            if key in self.tokens:
                if self.tokens[key] >= 1:
                    self.tokens[key] -= 1
                    self.add_admission_num(key)
                    return True
            else:
                self.add_admission_num(key)
                return True
        return False

    def add_admission_num(self, key):
        if key in self.admission_num:
            self.admission_num[key] += 1

File: src/main/java/org/serviceProvider/capacities/LeakyBucketForGateway.java
package org.serviceProvider.capacities;

import javafx.util.Pair;
import org.cloudbus.cloudsim.Log;
import org.cloudbus.cloudsim.core.CloudSim;
import org.enduser.networkPacket.NetworkPacket;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.utils.PolicyConstants;

import java.io.FileWriter;
import java.io.IOException;
import java.util.Map;

public class LeakyBucketForGateway implements LoadAdmission {
    private static final Logger logger = LoggerFactory.getLogger(LeakyBucketForGateway.class);

    int deviceId;

    /* for algorithm */
    double lastTime;

    Map<Pair<Integer, Integer>, Double> tokens; // Pair<Integer,Integer>: <serviceChainId, userLevel>

    Map<Pair<Integer, Integer>, Integer> capacities;

    Map<Pair<Integer, Integer>, Integer> admissionRate;

    /* for the record */

    int lastMinute; // Record once a minute.

    String leakyBucketHistoryFile; // for LeakyBucketExecutor history

    Map<Pair<Integer, Integer>, Integer> admissionNum; //serviceChainId, userLevel -> num

    public LeakyBucketForGateway(int deviceId, Map<Pair<Integer, Integer>, Double> tokens, Map<Pair<Integer, Integer>, Integer> capacities, Map<Pair<Integer, Integer>, Integer> admissionRate, String leakyBucketHistoryFile) {
        this.deviceId = deviceId;

        this.tokens = tokens;
        this.capacities = capacities;
        this.admissionRate = admissionRate;
        this.leakyBucketHistoryFile = leakyBucketHistoryFile;

        lastMinute = 0;

        lastTime = 0.0;

        admissionNum = capacities;
        initAdmissionNum();
    }

    /* return true or false */
    @Override
    public boolean isAdmission(NetworkPacket networkPacket) {

        if (networkPacket.destinationServiceId != 0) { // load admission only for gateway
            return true;
        } else {// serviceId = 0 equals service = "gateway";
            int thisMinute = (int) (CloudSim.clock() / PolicyConstants.aMinute);
            if (thisMinute != lastMinute) {
                //getLoadAdmissionExecutorHistories().add(loadAdmissionExecutorHistory);
                // Write directly to file.
                writeAdmissionExecutorResult(thisMinute);
                initAdmissionNum();
                lastMinute = thisMinute;
            }

            updateTokens();
            int serviceChainId = networkPacket.getServiceChainInfo().getServiceChainId();
            int userLevel = networkPacket.getEndUserInfo().getUserLevel();
            Pair<Integer, Integer> key = new Pair<>(serviceChainId, userLevel);

            if (tokens.containsKey(key)) {
                double t = tokens.get(key);
                if (t >= 1) {
                    tokens.put(key, t - 1);
                    addAdmissionNum(key);
                    return true;
                }
            } else {
                addAdmissionNum(key);
                return true; // If the threshold is not found, it is received directly.
            }
        }
        return false;
    }

    public void initAdmissionNum() {
        admissionNum.replaceAll((k, v) -> 0);
    }

    // write Result to the file
    public void writeAdmissionExecutorResult(int thisMinute) {
        FileWriter fileWriter = null;

        try {
            fileWriter = new FileWriter(leakyBucketHistoryFile, true);
            //fileWriter.append('\n');
            // time,serviceChainId,User levelId,requestNum,admissionNum
            for (Pair<Integer, Integer> key : admissionNum.keySet()) {
                fileWriter.append(String.valueOf(thisMinute));
                fileWriter.append(',');
                fileWriter.append(String.valueOf(deviceId));
                fileWriter.append(',');
                fileWriter.append(String.valueOf(key.getKey())); // serviceChainId,
                fileWriter.append(',');
                fileWriter.append(String.valueOf(key.getValue())); // userLevel
                fileWriter.append(',');
                fileWriter.append(String.valueOf(admissionNum.get(key)));
                fileWriter.append('\n');
            }

        } catch (Exception e) {
            System.out.println("Error in CsvFileWriter !!!");
            logger.error("Error in CsvFileWriter !!!");
        }
        try {
            if (fileWriter != null) {
                fileWriter.close();
            }
        } catch (IOException e) {
            logger.error("Error while flushing/closing fileWriter !!!");
        }
    }

    public void updateTokens() {

        for (Pair<Integer, Integer> key : tokens.keySet()) {
            if (admissionRate.containsKey(key) && capacities.containsKey(key)) {
                // Rate is the number of tokens placed per minute.
                double updateToken = min(capacities.get(key), tokens.get(key) + (CloudSim.clock() - lastTime) / PolicyConstants.aMinute * admissionRate.get(key));
                tokens.put(key, updateToken);
                //lastTime = CloudSim.clock();
            } else {
                Log.printLine("error- Cannot find keys for tokens or capabilities.");
            }
        }
        lastTime = CloudSim.clock();
    }

    public double min(double a, double b) {
        return Math.min(a, b);
    }

    public void addAdmissionNum(Pair<Integer, Integer> key) {
        int num;
        if (admissionNum.containsKey(key)) {
            num = admissionNum.get(key) + 1;
            admissionNum.put(key, num);
        }
    }


    /* getter and setter */
    public int getDeviceId() {
        return deviceId;
    }

    public void setDeviceId(int deviceId) {
        this.deviceId = deviceId;
    }

    public double getLastTime() {
        return lastTime;
    }

    public void setLastTime(double lastTime) {
        this.lastTime = lastTime;
    }

    public Map<Pair<Integer, Integer>, Double> getTokens() {
        return tokens;
    }

    public void setTokens(Map<Pair<Integer, Integer>, Double> tokens) {
        this.tokens = tokens;
    }

    public Map<Pair<Integer, Integer>, Integer> getCapacities() {
        return capacities;
    }

    public void setCapacities(Map<Pair<Integer, Integer>, Integer> capacities) {
        this.capacities = capacities;
    }

    public Map<Pair<Integer, Integer>, Integer> getAdmissionRate() {
        return admissionRate;
    }

    public void setAdmissionRate(Map<Pair<Integer, Integer>, Integer> admissionRate) {
        this.admissionRate = admissionRate;
    }

    public int getLastMinute() {
        return lastMinute;
    }

    public void setLastMinute(int lastMinute) {
        this.lastMinute = lastMinute;
    }

    public String getLeakyBucketHistoryFile() {
        return leakyBucketHistoryFile;
    }

    public void setLeakyBucketHistoryFile(String leakyBucketHistoryFile) {
        this.leakyBucketHistoryFile = leakyBucketHistoryFile;
    }

    public Map<Pair<Integer, Integer>, Integer> getAdmissionNum() {
        return admissionNum;
    }

    public void setAdmissionNum(Map<Pair<Integer, Integer>, Integer> admissionNum) {
        this.admissionNum = admissionNum;
    }
}
